### PR TITLE
[dynamic-igemm][fwd-xdlops]Assembly implementation of dynamic implicit-gemm generic tensor contraction for convolution-forward

### PR DESCRIPTION
**Pull request by fronteer**
1) Reduce implicit-gemm conv-forward kernel compile-time by
    * coding  in assembly
    * using dynamic parameters to reduce the number of kernel instances (all parameters of conv-descriptor are passed in run-time)
2) Support the same range of user cases as rocm-3.7's HIP-based implicit-gemm kernels (except group conv)
3) Take advantage of the gfx908 xdlops ISAs for accelerated performance (compared with gfx906)

Performance data form and validated network scope can be accessed from 

https://amdcloud-my.sharepoint.com/:x:/r/personal/qfzhang_amd_com/_layouts/15/doc2.aspx?sourcedoc=%7Bac04805b-8434-4eb7-a3ea-0a761133e308%7D&action=edit&activeCell=%27xdlops-fwd-09-23%27!R132&wdrcid=ef07b42e-2018-4641-9c72-2ef66f6f30ce&wdrldc=1&cid=5fb9804b-8a98-4f35-b6f4-249ecf9f9ab9

To test:
 #> make test_conv_igemm_dynamic_xdlops        or  whatever scripts calling "MIOpenDriver conv xxx"

All kernel is in `src/kernels/dynamic_igemm/igemm_gtc_fwd/` folder
- igemm_fwd_gtc_gfx908.s
- igemm_fwd_gtc_gfx908_004x064.inc
- igemm_fwd_gtc_gfx908_008x064.inc
- igemm_fwd_gtc_gfx908_016x016.inc
- igemm_fwd_gtc_gfx908_016x032.inc
- igemm_fwd_gtc_gfx908_016x064.inc
- igemm_fwd_gtc_gfx908_032x016.inc
- igemm_fwd_gtc_gfx908_032x032.inc
- igemm_fwd_gtc_gfx908_032x064.inc
- igemm_fwd_gtc_gfx908_032x128.inc
- igemm_fwd_gtc_gfx908_032x256.inc
- igemm_fwd_gtc_gfx908_064x004.inc
- igemm_fwd_gtc_gfx908_064x008.inc
- igemm_fwd_gtc_gfx908_064x016.inc
- igemm_fwd_gtc_gfx908_064x032.inc
- igemm_fwd_gtc_gfx908_064x064.inc
- igemm_fwd_gtc_gfx908_064x128.inc
- igemm_fwd_gtc_gfx908_064x256.inc
- igemm_fwd_gtc_gfx908_128x032.inc
- igemm_fwd_gtc_gfx908_128x064.inc
- igemm_fwd_gtc_gfx908_128x128.inc
- igemm_fwd_gtc_gfx908_128x256.inc
- igemm_fwd_gtc_gfx908_256x032.inc
- igemm_fwd_gtc_gfx908_256x064.inc
- igemm_fwd_gtc_gfx908_256x128.inc

perf data can be found in:
https://github.com/AMDComputeLibraries/MLOpen/issues/2535

### CODE REVIEW

**Review by asroy — DISMISSED** [review]

File: src/include/miopen/solver.hpp [hunk 1]
```diff
@@ -1204,6 +1204,77 @@ struct ConvAsmImplicitGemmV4R1DynamicBwd : SolverBase<ConvolutionContext>
...
+        const auto WaveSize = 64;
+        auto block_size     = (gemm_m_per_block / (wave_tile_m * wave_step_m * wave_repeat_m)) *
+                          (gemm_n_per_block / (wave_tile_n * wave_step_n * wave_repeat_n)) *
+                          WaveSize;
+        return block_size;
+    }
+
+    std::string GetKernelName()
+    {
+        std::string kernel_name = std::string("igemm_");
```
Comment (DrizztDoUrden) [inline src/include/miopen/solver.hpp:1242]: Use `std::ostringstream` here, please.

**Review by DrizztDoUrden — CHANGES_REQUESTED** [review]
Codecov really helps to review tbh. It's much easier to focus on the code when you can only see ten lines on a 4k monitor. Lucky me, it's not vertical.

**Review by carlushuang — APPROVED** [review]
LGTM

File: src/include/miopen/solver.hpp (hunk 1 above)
Comment (carlushuang) [reply src/include/miopen/solver.hpp]: @DrizztDoUrden this is change, and use the same code as in https://github.com/ROCmSoftwarePlatform/MIOpen/pull/442 (merged PR).

**Pull request merged**